python/pyproject.toml

```diff
@@ -18,7 +18,12 @@
 [build-system]
 requires = [
     "cython >= 0.29.31",
-    "oldest-supported-numpy>=0.14",
+    # Starting with NumPy 1.25, NumPy is (by default) as far back compatible
+    # as oldest-support-numpy was (customizable with a NPY_TARGET_VERSION
```

Comment:
Will this always be the case?

Reply:
What do you mean exactly with "always"? For all Python versions, or all numpy versions?  Or the full API?

It's a guarantee that is given by NumPy starting with version 1.25: https://numpy.org/devdocs/release/1.25.0-notes.html#compiling-against-the-numpy-c-api-is-now-backwards-compatible-by-default

(see also some details I posted on the parent issue: https://github.com/apache/arrow/issues/39532#issuecomment-1883045140)

Reply:
I mean that oldest-supported-numpy also accounts for bugs that cropped up in the past, e.g. broken ARM compatibility.
https://github.com/scipy/oldest-supported-numpy/blob/main/setup.cfg

Reply:
As far as I understand, oldest-supported-numpy pins to a _newer_ numpy version for some platforms / architectures (like arm) because the older version that is theoretically supported has bugs. But so that means that if we always use the latest available version for a platform, that should be fine? 

Of course, there might be a new bug occurring in the latest version, and at that point, defaulting to "the latest available version" might temporarily give issues, compared to pinning to a slightly older but "known to be OK" version (for example, if we know 1.25 is fine, we could pin to that in case the newest 1.26.x would introduce a bug). 
But that's a problem for all packages building against numpy, and so it would be good if numpy would have guidelines around this. And my understanding is that it is now recommended to built against the latest numpy version (although https://numpy.org/devdocs/dev/depending_on_numpy.html#build-time-dependency is not explicit about that, it just mentions that starting from 1.25, you can use that version to built packages that are also compatible with older numpy)

cc @seberg

Reply:
That sounds all exactly right.

I suppose the docs linked could recommend to compile against the latest available version (unless there be bugs).  This is required for major releases, i.e. NumPy 2 (you cannot build releases against it yet since there is no `rc`, though.  I still need to opaquify the descriptor struct a bit, which you may notice).

To continue supporting building against pre 2.0 releases, you may have to vendor `(numpy/_core/include/)numpy/npy_2_compat.h`.